### PR TITLE
Mkirk/webrtc m67

### PR DESCRIPTION
**Pull request by michaelkirk-signal**
PTAL @charlesmchen

### CODE REVIEW

File: Signal/src/call/PeerConnectionClient.swift [hunk 1]
```diff
@@ -393,30 +400,42 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
...
 
             DispatchQueue.main.async(execute: completion)
         }
     }
 
-    // MARK: Audio
+    // MARK: VideoCaptureSettingsDelegate
+
+    // MJK: fixme
+    var videoWidth: Int32 {
```
Comment (michaelkirk-signal) [inline Signal/src/call/PeerConnectionClient.swift]: Default width/height is TBD, but if you'd like to review the remaining changes, I could let you know which values I settle on.

File: Signal/src/call/PeerConnectionClient.swift [hunk 2]
```diff
@@ -331,49 +336,38 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
...
-        let cameraConstraints = RTCMediaConstraints(mandatoryConstraints: nil,
-                                                    optionalConstraints: nil)
-
-        // TODO: Revisit the cameraConstraints.
-        let videoSource = factory.avFoundationVideoSource(with: cameraConstraints)
-        self.localVideoSource = videoSource
-
-        self.videoCaptureSession = videoSource.captureSession
-        videoSource.useBackCamera = false
+        let videoSource = factory.videoSource()
```
Comment (michaelkirk-signal) [inline Signal/src/call/PeerConnectionClient.swift]: These are the primary upstream API changes, the fallout of which is this PR.

File: Signal/src/call/PeerConnectionClient.swift [hunk 3]
```diff
@@ -1092,6 +1110,90 @@ class HardenedRTCSessionDescription {
     }
 }
 
+protocol VideoCaptureSettingsDelegate: class {
+    var videoWidth: Int32 { get }
+    var videoHeight: Int32 { get }
+}
+
+class VideoCaptureController {
```
Comment (michaelkirk-signal) [inline Signal/src/call/PeerConnectionClient.swift]: This is based on the example class from upstream: https://webrtc.googlesource.com/src/+/c522e758f6a364c23e7232d4f402713601f0b35e%5E%21/

File: Signal/src/call/PeerConnectionClient.swift [hunk 4]
```diff
@@ -1092,6 +1110,90 @@ class HardenedRTCSessionDescription {
...
+            return
+        }
+
+        let fps = self.framesPerSecond(format: format)
+
+        capturer.startCapture(with: device, format: format, fps: fps)
+    }
+
+    public func stopCapture() {
+        self.capturer.stopCapture()
```
Comment (charlesmchen-signal) [inline Signal/src/call/PeerConnectionClient.swift]: why all of this unnecessary `self.`?  NBD.

File: Signal/src/call/PeerConnectionClient.swift [hunk 5]
```diff
@@ -1092,6 +1110,90 @@ class HardenedRTCSessionDescription {
...
+protocol VideoCaptureSettingsDelegate: class {
+    var videoWidth: Int32 { get }
+    var videoHeight: Int32 { get }
+}
+
+class VideoCaptureController {
+
+    let capturer: RTCCameraVideoCapturer
+    weak var settingsDelegate: VideoCaptureSettingsDelegate?
+    var isUsingFrontCamera: Bool = true
```
Comment (charlesmchen-signal) [inline Signal/src/call/PeerConnectionClient.swift]: I'm slightly concerned about thread safety.  We seem to access `VideoCaptureController` on the main thread and the "signaling queue".  `isUsingFrontCamera` for example is mutated and accessed on both, right?  Should we make the properties of this class private and synchronize access to them?

File: Signal/src/call/PeerConnectionClient.swift (hunk 4 above)
Comment (michaelkirk-signal) [reply Signal/src/call/PeerConnectionClient.swift]: I'm personally not a fan of the "implicit self", as I find it makes it harder to distinguish from a local variable. 

I would prefer that we always add the explicit self, even when not required. But I don't feel that strongly about it and am happy to conform if you have a preference for the implicit self.

File: Signal/src/call/PeerConnectionClient.swift (hunk 5 above)
Comment (michaelkirk-signal) [reply Signal/src/call/PeerConnectionClient.swift]: Good point - I've serialized read/write to the affected methods.

File: Signal/src/call/PeerConnectionClient.swift (hunk 1 above)
Comment (michaelkirk-signal) [reply Signal/src/call/PeerConnectionClient.swift]: After some testing, this actually seems like a reasonable default. I'm checking with Android to make sure we're consistent.

File: Signal/src/call/PeerConnectionClient.swift [hunk 6]
```diff
@@ -1092,6 +1109,121 @@ class HardenedRTCSessionDescription {
...
+    var videoWidth: Int32 { get }
+    var videoHeight: Int32 { get }
+}
+
+class VideoCaptureController {
+
+    let serialQueue = DispatchQueue(label: "org.signal.videoCaptureController")
+    let capturer: RTCCameraVideoCapturer
+    weak var settingsDelegate: VideoCaptureSettingsDelegate?
+    var isUsingFrontCamera: Bool = true
```
Comment (charlesmchen-signal) [inline Signal/src/call/PeerConnectionClient.swift]: private?

**Comment by michaelkirk-signal** [comment]
merged into release/2.27.0 with 89402d0d2, and followups in #3601

File: Signal/src/call/PeerConnectionClient.swift (hunk 6 above)
Comment (michaelkirk-signal) [reply Signal/src/call/PeerConnectionClient.swift]: Good point, done!

**Pull request closed without merging**